model/note/open.go

```diff
@@ -0,0 +1,140 @@
+package note
+
+import (
+	"net/http"
+	"net/url"
+
+	"github.com/cozy/cozy-stack/client/request"
+	"github.com/cozy/cozy-stack/model/instance"
+	"github.com/cozy/cozy-stack/model/sharing"
+	"github.com/cozy/cozy-stack/model/vfs"
+	"github.com/cozy/cozy-stack/pkg/config/config"
+	"github.com/cozy/cozy-stack/pkg/consts"
+	"github.com/cozy/cozy-stack/pkg/couchdb"
+	"github.com/cozy/cozy-stack/pkg/jsonapi"
+)
+
+type apiNoteURL struct {
+	DocID      string `json:"_id,omitempty"`
+	NoteID     string `json:"note_id"`
+	Subdomain  string `json:"subdomain"`
+	Instance   string `json:"instance"`
+	Sharecode  string `json:"sharecode,omitempty"`
+	PublicName string `json:"public_name,omitempty"`
+}
+
+func (n *apiNoteURL) ID() string                             { return n.DocID }
+func (n *apiNoteURL) Rev() string                            { return "" }
+func (n *apiNoteURL) DocType() string                        { return consts.NotesURL }
+func (n *apiNoteURL) Clone() couchdb.Doc                     { cloned := *n; return &cloned }
+func (n *apiNoteURL) SetID(id string)                        { n.DocID = id }
+func (n *apiNoteURL) SetRev(rev string)                      {}
+func (n *apiNoteURL) Relationships() jsonapi.RelationshipMap { return nil }
+func (n *apiNoteURL) Included() []jsonapi.Object             { return nil }
+func (n *apiNoteURL) Links() *jsonapi.LinksList              { return nil }
+func (n *apiNoteURL) Fetch(field string) []string            { return nil }
+
+// Open returns the parameters to create the URL where the note can be opened.
+func Open(inst *instance.Instance, file *vfs.FileDoc, code string) (*apiNoteURL, error) {
+	// Check that the file is a note
+	if _, err := fromMetadata(file); err != nil {
+		return nil, err
+	}
+
+	// Looks if the note is shared
+	fileID := file.ID()
+	sharing, err := getSharing(inst, fileID)
+	if err != nil {
+		return nil, err
+	}
+
+	var doc *apiNoteURL
+	if sharing == nil || sharing.Owner {
+		doc = openLocalNote(inst, fileID, code)
+	} else {
+		doc, err = openSharedNote(inst, sharing, fileID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Enforce DocID and PublicName with local values
+	doc.DocID = fileID
```

Comment:
Shouldn't you use `doc.SetID(fileID)` here?

Reply:
We don't care. The `SetID` is used by packages like `couchdb` and `jsonapi` that needs a generic way to manipulate a lot of doctypes (well, their types in go). But, when we are accessing/setting an ID for a struct in the same module, we often prefer the direct way to just assign it.